src/java/org/apache/cassandra/cql3/statements/schema/CreateTableStatement.java

```diff
@@ -118,6 +120,34 @@ public Keyspaces apply(Keyspaces schema)
         return schema.withAddedOrUpdated(keyspace.withSwapped(keyspace.tables.with(table)));
     }
 
+    @Override
+    public void validate(ClientState state)
+    {
+        super.validate(state);
+
+        // since there are multiple guardrails guarding this, we first check that guardrails are globally enabled as an
+        // optimization for the case where they are disabled, so we don't have to do the same check on every guardrail
+        if (Guardrails.enabled(state))
+        {
+            // Guardrails on table properties
+            Guardrails.allowedTableProperties.guard(attrs.updatedProperties(), attrs::removeProperty, state);
+
+            // Guardrail on columns per table
+            Guardrails.columnsPerTable.guard(rawColumns.size(), tableName, false, state);
+
+            // Guardrails on number of tables
+            if (Guardrails.tablesLimit.enabled(state))
+            {
+                int totalUserTables = Schema.instance.getUserKeyspaces()
+                                                     .stream()
+                                                     .map(Keyspace::open)
+                                                     .mapToInt(keyspace -> keyspace.getColumnFamilyStores().size())
+                                                     .sum();
+                Guardrails.tablesLimit.guard(totalUserTables + 1, tableName, false, state);
+            }
+        }
+    }
```

Comment:
with the new API there is potential for optimization; also good to make sure config doesn't change between calls

```
var conf = Guardrails.getOrCreate(state);
if (Guardrails.enabled(conf, state))
{
  Guardrails.allowedTableProperties.guard(conf, attrs.updatedProperties(), attrs::removeProperty, state);
  Guardrails.columnsPerTable.guard(conf, rawColumns.size(), tableName, false, state);
if (Guardrails.tablesLimit.enabled(conf, state))
              {
                  int totalUserTables = Schema.instance.getUserKeyspaces()
                                                       .stream()
                                                       .map(Keyspace::open)
                                                       .mapToInt(keyspace -> keyspace.getColumnFamilyStores().size())
                                                       .sum();
                  Guardrails.tablesLimit.guard(conf, totalUserTables + 1, tableName, false, state);
              }
}
```

I am cool not doing in this patch; but we do this in track warnings so queries see a consistent config

Reply:
I don't think that's going to guarantee that all guardrails see a consistent config per se since `GuardrailsConfig` is mutable. I understand that `Guardrails.getOrCreate` would need to return an immutable copy of `GuardrailsConfig`, wouldn't it? Nevertheless I don't see an immediate need to provide a consistent view of the guardrails config.

Reply:
currently we are mutating in place, so the default implementation doesn't really do anything.  I was mostly thinking about your heiractical patch.

100% cool not doing for now. This topic will be brought up when track_warnings gets integrated as we did want consistency for the query